metrics/metrics.go

```diff
@@ -0,0 +1,150 @@
+		},
+
+		[]string{
+			// The HTTP status code AWS returned
+			"code",
+			// The arn of the IAM role being requested
+			"role_arn",
+		},
+	)
+
+	// IamRequestSec tracks timing of IAM requests.
+	IamRequestSec = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: namespace,
+			Subsystem: "iam",
+			Name:      "request_duration_seconds",
+			Help:      "Time taken to complete IAM request in seconds.",
+
+			Buckets: prometheus.ExponentialBuckets(0.001, 2, 14),
+		},
+		[]string{
+			// The HTTP status code AWS returned
+			"code",
+			// The arn of the IAM role being requested
+			"role_arn",
+		},
+	)
+
+	// HTTPRequestCount counts number of HTTP requests served split by handler.
+	HTTPRequestCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: namespace,
+			Subsystem: "http",
+			Name:      "requests_total",
+			Help:      "Number of http requests served by kube2iam.",
+		},
+		[]string{
+			// The HTTP status code kube2iam returned
+			"code",
+			// The HTTP method being served
+			"method",
+			// The name of the handler being served
+			"handler",
+		},
+	)
+
+	// HTTPRequestSec tracks timing of served HTTP requests.
+	HTTPRequestSec = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: namespace,
+			Subsystem: "http",
+			Name:      "request_duration_seconds",
+			Help:      "Time taken for kube2iam to serve HTTP request.",
+
+			Buckets: prometheus.ExponentialBuckets(0.001, 2, 14),
+		},
+		[]string{
+			// The HTTP status code kube2iam returned
+			"code",
+			// The HTTP method being served
+			"method",
+			// The name of the handler being served
+			"handler",
+		},
+	)
+
+	// HealthcheckStatus reports the current healthcheck status of kube2iam.
+	HealthcheckStatus = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: namespace,
+			Subsystem: "healthcheck",
+			Name:      "status",
+			Help:      "The healthcheck status at scrape time. A value of 1 means it is passing, 0 means it is failing.",
+		},
+	)
+
+	// Info reports various static information about the running kube2iam binary.
+	Info = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "info",
+			Help:      "Informational labels about the kube2iam process.",
+		},
+		[]string{
+			// The version of kube2iam running
+			"version",
+			// The build date of the kube2iam version
+			"build_date",
+			// The commit hash of the kube2iam version
+			"commit_hash",
+		},
+	)
+)
+
+type lvsProducer func() []string
+
+// Init ensures prometheus knows about all the metrics tracked by kube2iam and initializes the values with default
+// label sets.
+func Init() {
+	prometheus.MustRegister(IamRequestCount)
+	prometheus.MustRegister(IamRequestSec)
+	prometheus.MustRegister(HTTPRequestCount)
+	prometheus.MustRegister(HTTPRequestSec)
+	prometheus.MustRegister(HealthcheckStatus)
+	prometheus.MustRegister(Info)
+
+	for _, val := range []string{"Success", "UnknownError"} {
```

Comment:
Why do we bother setting these default values?

Reply:
I was reading that it's a good practice to initialize the time series for any known label values to avoid data missing errors when querying.

Reply:
I've resolved to still keep this initialization here (only metrics without labels are initialized) but I've removed all the HTTP ones from server.go.